File: ui/src/ipc.ts
declare global {
  interface Window {
    ipc?: { postMessage(message: string): void };
    __civReceiveEvents?: (events: string[]) => void;
  }
}

export type UiAction =
  | { action: "EndTurn" }
  | { action: "AddGold"; data: { player: number; amount: number } }
  | {
      action: "SpawnUnit";
      data: { owner: number; q: number; r: number; kind: string };
    }
  | {
      action: "CreateCity";
      data: { owner: number; q: number; r: number; name: string };
    }
  | { action: "RequestSnapshot" };

export async function dispatchCommand(action: UiAction): Promise<number> {
  if (!window.ipc) {
    throw new Error("UI overlay IPC is not available");
  }
  window.ipc.postMessage(JSON.stringify(action));
  return 0;
}

export function subscribeEvents(onEvents: (events: string[]) => void): () => void {
  const handler = (event: Event) => {
    const detail = (event as CustomEvent<string[]>).detail;
    if (Array.isArray(detail)) {
      onEvents(detail);
    }
  };
  window.addEventListener("civ:events", handler);
  return () => window.removeEventListener("civ:events", handler);
}

File: launcher/src/ipc.rs
use civ_commands::{GameCommand, UiCommand};
use civ_common::PlayerId;
use civ_hex::HexCoordinate;

use crate::commands::UiAction;

pub fn parse_command(action: UiAction) -> UiCommand {
    match action {
        UiAction::EndTurn => UiCommand::Dispatch(GameCommand::EndTurn),
        UiAction::AddGold { player, amount } => UiCommand::Dispatch(GameCommand::AddGold {
            player: PlayerId(player),
            amount,
        }),
        UiAction::SpawnUnit {
            owner,
            q,
            r,
            kind,
        } => UiCommand::Dispatch(GameCommand::SpawnUnit {
            owner: PlayerId(owner),
            at: HexCoordinate::new(q, r),
            kind,
        }),
        UiAction::CreateCity {
            owner,
            q,
            r,
            name,
        } => UiCommand::Dispatch(GameCommand::CreateCity {
            owner: PlayerId(owner),
            at: HexCoordinate::new(q, r),
            name,
        }),
        UiAction::RequestSnapshot => UiCommand::RequestStateSnapshot,
    }
}
